File: leetcode-solutions/python/213. House Robber II.py
# https://leetcode.com/problems/house-robber-ii/

class Solution:
    def rob(self, ns: List[int]) -> int:
        if len(ns) <= 3:
            return max(ns)

        def linear_rob(nums):
            dp = [nums[0]]
            for i in range(1, len(nums)):
                if i == 1:
                    dp.append(max(nums[i], dp[i - 1]))
                    continue
                dp.append(max(nums[i] + dp[i - 2], dp[i - 1]))
            return dp[-1]
        # pick first one or skip
        # if pick first one, have to skip last one
        # without first one, with first one
        return max(linear_rob(ns[1:]), linear_rob(ns[:-1]))

File: leetcode-solutions/ruby/213. House Robber II.rb
# https://leetcode.com/problems/house-robber-ii/

# @param {Integer[]} nums
# @return {Integer}
def rob(nums)
    return nums.max if nums.size <= 2
    
    [max_rob_amount(nums[1..-1]), max_rob_amount(nums[0...-1])].max
end

def max_rob_amount(nums)
    max_amount_at = Array.new(nums.size+2){0}
    right_index = nums.size - 1
    right_index.downto(0) do |i|
        max_amount_at[i] = [nums[i] + max_amount_at[i+2], max_amount_at[i+1]].max
    end
    max_amount_at.first
end